src/Compilers/CSharp/Test/Semantic/Semantics/NullCheckedParameterTests.cs

```diff
@@ -170,5 +170,295 @@ class C
                     //     public static extern int M(int x!);
                     Diagnostic(ErrorCode.ERR_MustNullCheckInImplementation, "x").WithArguments("x").WithLocation(6, 36));
         }
+
+        [Fact]
+        public void NullCheckedMethodDeclaration()
+        {
+            var source = @"
+class C
+{
+    void M(string name!) { }
+}";
+            CreateCompilation(source).VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void FailingNullCheckedArgList()
+        {
+            var source = @"
+class C
+{
+    void M()
+    {
+        M2(__arglist(1, 'M'));
+    }
+    void M2(__arglist!)
+    {
+    }
+}";
+            CreateCompilation(source).VerifyDiagnostics( 
+                    // (8,22): error CS1003: Syntax error, ',' expected
+                    //     void M2(__arglist!)
+                    Diagnostic(ErrorCode.ERR_SyntaxError, "!").WithArguments(",", "!").WithLocation(8, 22));
+        }
+
+        [Fact]
+        public void NullCheckedArgList()
+        {
+            var source = @"
+class C
+{
+    void M()
+    {
+        M2(__arglist(1!, 'M'));
+    }
+    void M2(__arglist)
+    {
+    }
+}";
+            CreateCompilation(source).VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void NullCheckedMethodValidationWithOptionalParameter()
+        {
+            var source = @"
+class C
+{
+    void M(string name! = null) { }
+}";
+            CreateCompilation(source).VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void NullCheckedMethodValidationWithOptionalParameterNoSpaces()
+        {
+            var source = @"
+class C
+{
+    void M(string name!=null) { }
```

Comment:
Some of these are parser specific and don't really need to be repeated in the semantic tests. No harm in having them though.

Reply:
Can get rid of the spacing ones if you prefer.